tests/std/tests/P0896R4_ranges_alg_is_permutation/test.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+//
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
+struct instantiator {
+    using P = pair<int, int>;
+
+    static constexpr array<P, 8> pairs = {
+        P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}};
+    static constexpr array<int, 8> not_pairs = {0, 0, 0, 0, 1, 1, 1, 1};
+
+    template <class Fwd1, class Fwd2>
+    static constexpr void call() {
+        const Fwd1 range1{pairs};
+        const Fwd2 range2{not_pairs};
+
+        // defaulted predicate and projections
+        assert(ranges::is_permutation(range1, range1));
+        assert(ranges::is_permutation(
+            ranges::begin(range2), ranges::end(range2), ranges::begin(range2), ranges::end(range2)));
+
+        // explicit predicate
+        assert(ranges::is_permutation(range2, range2, ranges::equal_to{}));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range1),
+            ranges::end(range1), ranges::equal_to{}));
+
+        // explicit predicate and one projection
+        assert(ranges::is_permutation(range1, range2, ranges::equal_to{}, get_first));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_first));
+
+        // explicit predicate and two projections
+        assert(ranges::is_permutation(range1, range2, ranges::equal_to{}, get_first, identity{}));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_first, identity{}));
+
+        // negative case
+        assert(!ranges::is_permutation(range1, range2, ranges::equal_to{}, get_second, identity{}));
+        assert(!ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_second, identity{}));
+    }
+};
+
+#ifdef TEST_EVERYTHING
+int main() {
+    // No constepxr test here; this test_fwd_fwd call exceeds the maximum number of steps in a constexpr computation
+    test_fwd_fwd<instantiator, const pair<int, int>, const int>();
+}
+#else // ^^^ test all range combinations // test only interesting range combos vvv
+

```

Comment:
Nitpick: do we need this blank line?  In poking around how we use `#else`, I don't see a blank line after most occurrences.